build_runner/lib/src/package_graph/apply_builders.dart

```diff
@@ -245,11 +249,23 @@ Future<List<BuildPhase>> createBuildPhases(
     bool isReleaseMode) async {
   validateBuilderConfig(builderApplications, targetGraph.rootPackageConfig,
       builderConfigOverrides, _logger);
-  final cycles = stronglyConnectedComponents<String, TargetNode>(
-      targetGraph.allModules.values,
-      (node) => node.target.key,
-      (node) =>
-          node.target.dependencies?.map((key) => targetGraph.allModules[key]));
+  List<List<TargetNode>> cycles;
+  try {
+    cycles = stronglyConnectedComponents<String, TargetNode>(
+        targetGraph.allModules.values,
+        (node) => node.target.key,
+        (node) => node.target.dependencies?.map((key) {
+              if (!targetGraph.allModules.containsKey(key)) {
+                _logger
+                    .severe('${node.target.key} declares a dependency on $key '
+                        'but it does not exist');
+                throw const _MissingDependency();
```

Comment:
In some ways I think it would be cleaner to simply throw here, and add more information to the exception so that it can have a nice toString().

Then, we can catch the error at a higher level and not have to return an empty list which doesn't really seem like the correct behavior (it happens to end up fine but you get the weird extra log about there being nothing to build)

Reply:
yeah I already added the `CannotBuildException` in another recent PR which is handled well.
I'm not sure why I didn't think to use it here 😝